code/android-core-library/src/phone/java/com/adobe/marketing/mobile/services/caching/FileCacheService.java

```diff
@@ -0,0 +1,126 @@
+/*
+  Copyright 2022 Adobe. All rights reserved.
+  This file is licensed to you under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License. You may obtain a copy
+  of the License at http://www.apache.org/licenses/LICENSE-2.0
+  Unless required by applicable law or agreed to in writing, software distributed under
+  the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+  OF ANY KIND, either express or implied. See the License for the specific language
+  governing permissions and limitations under the License.
+ */
+
+package com.adobe.marketing.mobile.services.caching;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import com.adobe.marketing.mobile.services.DeviceInforming;
+import com.adobe.marketing.mobile.services.Log;
+import com.adobe.marketing.mobile.services.ServiceConstants;
+import com.adobe.marketing.mobile.util.StringUtils;
+
+import java.io.File;
+import java.util.Date;
+import java.util.Map;
+
+/**
+ * A {@code CacheService} that uses file storage as the cache store.
+ */
+public class FileCacheService implements CacheService {
+    private static final String TAG = "FileCacheService";
+    static final String ROOT_CACHE_DIR_NAME = "aepsdkcache";
+
+    private final CacheFileManager cacheFileManager;
+
+    public FileCacheService(@NonNull final DeviceInforming deviceInfoService) {
+        this.cacheFileManager = new CacheFileManager(deviceInfoService, ROOT_CACHE_DIR_NAME);
+    }
+
+    /**
+     * Creates or updates the key-value pair in the cache.
+     * Stores the content and its metadata as separate files internally in the
+     * cache directory supplied by {@code DeviceInforming#getApplicationCacheDir}
+     *
+     * @param cacheName name of the bucket where the cache entry is to be created
+     * @param key       key for the cache entry
+     * @param value     the value that is to be associated with {@code key}
+     * @return true if the value for the key was created or updated; false otherwise.
+     */
+    @Override
+    public boolean set(@NonNull String cacheName, @NonNull String key, @NonNull CacheEntry value) {
+        if (!canProcess(cacheName, key)) return false;
```

Comment:
Can you validate `CacheEntry` here 
`value != null and value.getStream() != null and value.getExpiry() != null` so that it does not fail downstream?

Reply:
IMO, We should mark the parameters with NonNull/Nullable annotations and avoid redundant null and validations.

Reply:
The IDE only shows warnings when calling these APIs with null values from Java. I prefer failing early rather than some internal logic throwing a null pointer exception. 
Another approach would be to throw `NullPointerException` from the `CacheEntry` constructor when passing null values.